## 1. Import Libraries

In [39]:
import os

import numpy as np

import pandas as pd

from sklearn.model_selection import train_test_split

## 2. Read the Data

In [40]:
Project_Def=r"C:\Users\Debasish Das\Desktop\Flight_Price_Prediction"
Data_Dir="Dataset"

In [41]:
def get_data(name):
    file_name=f"{name}.csv"
    file_path=os.path.join(Project_Def,Data_Dir,file_name)
    return pd.read_csv(file_path)

In [42]:
flight=get_data("flight_price")
flight

Airline Date_of_Journey    Source Destination   
0           IndiGo      24/03/2019  Banglore   New Delhi  \
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops   
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop  \
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [43]:
flight.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


-  In the dataset `10683` rows and `11` Features
- `Route` and `Total_stops` features have some missing value
- some Features are not appropriate datatypes.

## 3. Preliminary Analysis

### 3.1 Check Data Types

In [44]:
flight.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

### 3.2 Check Duplicates

In [45]:
flight.duplicated().sum()

220

In [46]:
flight.loc[flight.duplicated(keep=False)].sort_values(["Date_of_Journey"])

Airline Date_of_Journey    Source Destination   
9848     Air India      01/03/2019  Banglore   New Delhi  \
6321     Air India      01/03/2019  Banglore   New Delhi   
7502      SpiceJet      03/03/2019  Banglore   New Delhi   
6196        IndiGo      03/03/2019  Banglore   New Delhi   
10275       IndiGo      03/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
1312   Jet Airways       9/06/2019     Delhi      Cochin   
9191   Jet Airways       9/06/2019     Delhi      Cochin   
9216   Jet Airways       9/06/2019     Delhi      Cochin   
8446   Jet Airways       9/06/2019     Delhi      Cochin   
8496   Jet Airways       9/06/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops   
9848   BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops  \
6321   BLR → BOM → AMD → DEL    08:50  23:55 02 Mar   39h 5m     2 stops   
7502               BLR → DEL    20:30         23:20   2h 50m    non-stop   
6196               BLR → DEL    08:30         11:30       3h    non-stop   
10275              BLR → DEL    08:30         11:30       3h    non-stop   
...                      ...      ...           ...      ...         ...   
1312   DEL → NAG → BOM → COK    06:45  12:35 10 Jun  29h 50m     2 stops   
9191   DEL → JAI → BOM → COK    09:40  12:35 10 Jun  26h 55m     2 stops   
9216   DEL → IDR → BOM → COK    05:30  04:25 10 Jun  22h 55m     2 stops   
8446   DEL → NAG → BOM → COK    06:45  12:35 10 Jun  29h 50m     2 stops   
8496   DEL → JAI → BOM → COK    09:40  12:35 10 Jun  26h 55m     2 stops   

                   Additional_Info  Price  
9848                       No info  17135  
6321                       No info  17135  
7502                       No info   6860  
6196                       No info   6860  
10275                      No info   6860  
...                            ...    ...  
1312                       No info  13376  
9191   In-flight meal not included  10368  
9216                       No info  13292  
8446                       No info  13376  
8496                       No info  13014  

[415 rows x 11 columns]

### 3.3 Observation

- The type of  `Date_of_Journey`, `Dep_Time`, and `Arrival_Time`  should be changed to `DateTime`
- In-flight Dataset 220 duplicates row . You should remove those rows.
- The type of `Duration` and `Total_Stops` is mixed. It should be numeric type.

## 4. Detailed Analysis

#### 1. Airline

In [47]:
flight.Airline.unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [48]:
(
    flight
    .Airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
)

0             Indigo
1          Air India
2        Jet Airways
3             Indigo
4             Indigo
            ...     
10678       Air Asia
10679      Air India
10680    Jet Airways
10681        Vistara
10682      Air India
Name: Airline, Length: 10683, dtype: object

- some of the values are inconsistent

#### 2. Date_of_Jour

In [49]:
pd.to_datetime(flight.Date_of_Journey,dayfirst=True)

0       2019-03-24
1       2019-05-01
2       2019-06-09
3       2019-05-12
4       2019-03-01
           ...    
10678   2019-04-09
10679   2019-04-27
10680   2019-04-27
10681   2019-03-01
10682   2019-05-09
Name: Date_of_Journey, Length: 10683, dtype: datetime64[ns]

#### 3. Source

In [50]:
flight.Source.unique()

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

#### 4. Destination

In [51]:
flight.Destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

#### 5. Dep_Time

In [52]:
flight.Dep_Time

0        22:20
1        05:50
2        09:25
3        18:05
4        16:50
         ...  
10678    19:55
10679    20:45
10680    08:20
10681    11:30
10682    10:55
Name: Dep_Time, Length: 10683, dtype: object

In [53]:
(
    flight
    .Dep_Time
    .loc[lambda ser: ser.str.contains("[^0-9:]")]
)

Series([], Name: Dep_Time, dtype: object)

In [54]:
pd.to_datetime(flight.Dep_Time).dt.time

C:\Users\Debasish Das\AppData\Local\Temp\ipykernel_9952\2899660464.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(flight.Dep_Time).dt.time


0        22:20:00
1        05:50:00
2        09:25:00
3        18:05:00
4        16:50:00
           ...   
10678    19:55:00
10679    20:45:00
10680    08:20:00
10681    11:30:00
10682    10:55:00
Name: Dep_Time, Length: 10683, dtype: object

#### 6. Arrival Time

In [55]:
flight.Arrival_Time.sample(5)

1548           20:20
5605    04:25 19 May
9867    12:00 25 May
8910           19:15
3760           19:40
Name: Arrival_Time, dtype: object

In [56]:
(
    flight.Arrival_Time
    .loc[ lambda ser : ser.str.contains("[^0-9:]")]
    .str.split(" ",n=1)
    .str.get(1)
    .unique()
)

array(['22 Mar', '10 Jun', '13 Mar', '02 Mar', '10 May', '04 Mar',
       '13 Jun', '28 May', '19 Mar', '07 May', '02 Jun', '16 Jun',
       '19 May', '16 May', '28 Jun', '02 May', '28 Mar', '19 Jun',
       '04 Apr', '25 Mar', '07 Mar', '25 Jun', '07 Jun', '25 May',
       '13 May', '16 Mar', '22 May', '10 Apr', '04 Jun', '20 May',
       '28 Apr', '25 Apr', '10 Mar', '19 Apr', '13 Apr', '02 Apr',
       '23 Mar', '22 Apr', '11 May', '07 Apr', '03 May', '08 Mar',
       '03 Mar', '05 Mar', '22 Jun', '04 May', '26 May', '16 Apr',
       '26 Jun', '29 May', '29 Jun', '29 Mar', '23 May', '17 Jun'],
      dtype=object)

#### 7. Duration

In [57]:
flight.Duration

0        2h 50m
1        7h 25m
2           19h
3        5h 25m
4        4h 45m
          ...  
10678    2h 30m
10679    2h 35m
10680        3h
10681    2h 40m
10682    8h 20m
Name: Duration, Length: 10683, dtype: object

In [58]:
(
    flight
    .Duration
    .loc[lambda ser : -ser.str.contains("m")]
)

2        19h
18       23h
33       22h
44       12h
53        3h
        ... 
10591    23h
10638    14h
10639    38h
10673    15h
10680     3h
Name: Duration, Length: 1031, dtype: object

In [59]:
(
    flight
    .Duration
    .loc[lambda ser : -ser.str.contains("h")]
)

6474    5m
Name: Duration, dtype: object

In [60]:
flight.iloc[[6474]]

Airline Date_of_Journey  Source Destination                  Route   
6474  Air India       6/03/2019  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD  \

     Dep_Time Arrival_Time Duration Total_Stops Additional_Info  Price  
6474    16:50        16:55       5m     2 stops         No info  17327

- The observation duration 5 minites .It is not a valid entry. We will delete this entry

In [61]:
(
    flight
    .Duration
    .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"],axis=1)
    .assign( hour = lambda df_ :(
        df_.hour.str.replace("h","")
        .astype(int)
        .mul(60)
    ),
    minute = lambda df_ :(
        df_.minute.str.replace("m","")
        .fillna(0)
        .astype(int)
        
    )
            
    )
    .sum(axis=1)
    #.dtypes
   #.isna().sum()         
)

0         170
1         445
2        1140
3         325
4         285
         ... 
10678     150
10679     155
10680     180
10681     160
10682     500
Length: 10682, dtype: int64

In [62]:
(
    flight
    .Duration
   .drop(index=6474)
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"],axis=1)
    .assign( hour = lambda df_ :(
        df_.hour.str.replace("h","")
        .astype(int)
        .mul(60)
    ),
    minute = lambda df_ :(
        df_.minute.str.replace("m","")
        .fillna(0) 
        .astype(int)
        
    )
            
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    #.dtypes
   #.isna().sum()         
)

duration_minutes
0                   170
1                   445
2                  1140
3                   325
4                   285
...                 ...
10678               150
10679               155
10680               180
10681               160
10682               500

[10682 rows x 1 columns]

#### 8. Total_stops

In [63]:
(
    flight.Total_Stops.unique()
)

array(['non-stop', '2 stops', '1 stop', '3 stops', nan, '4 stops'],
      dtype=object)

In [64]:
(
flight.Total_Stops
    .replace("non-stop","0")
    .str.replace(" stops?","",regex=True) # it is detect "stop" or "stops"
    #.astype(int)n # nan value is here 
    .pipe ( lambda ser : pd.to_numeric(ser))
    
)

0        0.0
1        2.0
2        2.0
3        1.0
4        1.0
        ... 
10678    0.0
10679    0.0
10680    0.0
10681    0.0
10682    2.0
Name: Total_Stops, Length: 10683, dtype: float64

#### 9. additional_info

In [65]:
flight.Additional_Info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class',
       'Red-eye flight', '2 Long layover'], dtype=object)

## 5. Cleaning Opration

In [66]:
def convert_minutes(ser):
    return (
        ser
    .str.split(" ",expand=True)
    .set_axis(["hour","minute"],axis=1)
    .assign( hour = lambda df_ :(
        df_.hour.str.replace("h","")
        .astype(int)
        .mul(60)
    ),
    minute = lambda df_ :(
        df_.minute.str.replace("m","")
        .fillna(0) 
        .astype(int)
        
    )
            
    )
    .sum(axis=1)
    .rename("duration_minutes")
    .to_frame()
    )
    

In [67]:
def clean_data(df):
    return(
        df
        .drop(index=6474)
        .drop_duplicates()
        .assign(**{
            col:df[col].str.strip()
            for col in df.select_dtypes(include="O").columns
        })
        .rename(columns=str.lower)
        .assign(
            airline=lambda df_:(
                df_
    .airline
    .str.replace(" Premium economy","")
    .str.replace(" Business","")
    .str.title()
            ),
            date_of_journey= lambda df_:pd.to_datetime(flight.Date_of_Journey,dayfirst=True),
            dep_time= lambda df_: pd.to_datetime(flight.Dep_Time).dt.time,
            arrival_time= lambda df_: pd.to_datetime(flight.Arrival_Time).dt.time,
            duration= lambda df_ : df_.duration.pipe(convert_minutes),
            total_stops= lambda df_ :(
				df_
				.total_stops
				.replace("non-stop", "0")
				.str.replace(" stops?", "", regex=True)
				.pipe(lambda ser: pd.to_numeric(ser))
			),
            additional_info= lambda df_ : df_.additional_info.replace("No info","No Info")
        )
        .drop(columns=["route"])
        )

In [68]:
clean_data(flight)

C:\Users\Debasish Das\AppData\Local\Temp\ipykernel_9952\1830914962.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time= lambda df_: pd.to_datetime(flight.Dep_Time).dt.time,
C:\Users\Debasish Das\AppData\Local\Temp\ipykernel_9952\1830914962.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time= lambda df_: pd.to_datetime(flight.Arrival_Time).dt.time,


airline date_of_journey    source destination  dep_time   
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00  \
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

In [69]:
flights_cleaned = clean_data(flight)
flights_cleaned

C:\Users\Debasish Das\AppData\Local\Temp\ipykernel_9952\1830914962.py:20: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  dep_time= lambda df_: pd.to_datetime(flight.Dep_Time).dt.time,
C:\Users\Debasish Das\AppData\Local\Temp\ipykernel_9952\1830914962.py:21: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  arrival_time= lambda df_: pd.to_datetime(flight.Arrival_Time).dt.time,


airline date_of_journey    source destination  dep_time   
0           Indigo      2019-03-24  Banglore   New Delhi  22:20:00  \
1        Air India      2019-05-01   Kolkata    Banglore  05:50:00   
2      Jet Airways      2019-06-09     Delhi      Cochin  09:25:00   
3           Indigo      2019-05-12   Kolkata    Banglore  18:05:00   
4           Indigo      2019-03-01  Banglore   New Delhi  16:50:00   
...            ...             ...       ...         ...       ...   
10678     Air Asia      2019-04-09   Kolkata    Banglore  19:55:00   
10679    Air India      2019-04-27   Kolkata    Banglore  20:45:00   
10680  Jet Airways      2019-04-27  Banglore       Delhi  08:20:00   
10681      Vistara      2019-03-01  Banglore   New Delhi  11:30:00   
10682    Air India      2019-05-09     Delhi      Cochin  10:55:00   

      arrival_time  duration  total_stops additional_info  price  
0         01:10:00       170          0.0         No Info   3897  
1         13:15:00       445          2.0         No Info   7662  
2         04:25:00      1140          2.0         No Info  13882  
3         23:30:00       325          1.0         No Info   6218  
4         21:35:00       285          1.0         No Info  13302  
...            ...       ...          ...             ...    ...  
10678     22:25:00       150          0.0         No Info   4107  
10679     23:20:00       155          0.0         No Info   4145  
10680     11:20:00       180          0.0         No Info   7229  
10681     14:10:00       160          0.0         No Info  12648  
10682     19:15:00       500          2.0         No Info  11753  

[10462 rows x 10 columns]

## 6.Split Data

In [70]:
flights_final = flights_cleaned.sample(10000)

In [71]:
X = flights_final.drop(columns="price")
y = flights_final.price.copy()

In [72]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(6400, 9) (6400,)
(1600, 9) (1600,)
(2000, 9) (2000,)


## Export The subset

In [73]:
def export_data(X, y, name):
	file_name = f"{name}.csv"
	file_path = os.path.join(Project_Def,Data_Dir,file_name)

	X.join(y).to_csv(file_path, index=False)

	return pd.read_csv(file_path).head()

In [74]:
export_data(X_train, y_train, "train")

airline date_of_journey   source destination  dep_time   
0  Multiple Carriers      2019-06-03    Delhi      Cochin  08:30:00  \
1           Spicejet      2019-05-24  Kolkata    Banglore  11:35:00   
2        Jet Airways      2019-06-27    Delhi      Cochin  13:25:00   
3          Air India      2019-06-12  Kolkata    Banglore  10:00:00   
4             Indigo      2019-06-24    Delhi      Cochin  16:10:00   

  arrival_time  duration  total_stops              additional_info  price  
0     19:00:00       630          1.0                      No Info  11200  
1     18:50:00       435          1.0                      No Info   8479  
2     19:00:00      1775          2.0  In-flight meal not included  11150  
3     13:45:00      1665          2.0                      No Info  12224  
4     22:30:00       380          1.0                      No Info   6412

In [75]:
export_data(X_val, y_val, "val")

airline date_of_journey    source destination  dep_time arrival_time   
0  Jet Airways      2019-03-06  Banglore   New Delhi  11:40:00     07:40:00  \
1    Air India      2019-03-18  Banglore   New Delhi  13:50:00     18:40:00   
2       Indigo      2019-03-21     Delhi      Cochin  07:35:00     16:10:00   
3  Jet Airways      2019-03-21     Delhi      Cochin  19:45:00     18:50:00   
4  Jet Airways      2019-05-01   Kolkata    Banglore  08:25:00     18:15:00   

   duration  total_stops              additional_info  price  
0      1200          1.0                      No Info  16736  
1      1730          1.0                      No Info  11161  
2       515          1.0                      No Info   6558  
3      1385          1.0  In-flight meal not included   5406  
4       590          1.0  In-flight meal not included  10844

In [76]:
export_data(X_test, y_test, "test")

airline date_of_journey source destination  dep_time   
0        Jet Airways      2019-03-27  Delhi      Cochin  19:15:00  \
1  Multiple Carriers      2019-06-06  Delhi      Cochin  09:45:00   
2        Jet Airways      2019-06-01  Delhi      Cochin  20:00:00   
3        Jet Airways      2019-03-21  Delhi      Cochin  20:50:00   
4  Multiple Carriers      2019-05-18  Delhi      Cochin  13:00:00   

  arrival_time  duration  total_stops              additional_info  price  
0     04:25:00       550          1.0  In-flight meal not included   6540  
1     22:30:00       765          1.0                      No Info   9646  
2     04:25:00       505          2.0  In-flight meal not included  15318  
3     04:25:00       455          1.0  In-flight meal not included   9212  
4     21:00:00       480          1.0                      No Info   9385